File: docs/observations/2026-09-06-astra-executable-specimen/specimen.js
// Prototype: validated archived specimens -> documentation. No request execution.
const fs = require('fs'), crypto = require('crypto'), util = require('util');
const hash = bytes => crypto.createHash('sha256').update(bytes).digest('hex');
const need = (ok, reason) => { if (!ok) throw Error(reason); };
function specimen(entry) {
  const raw = fs.readFileSync(entry.request), receipt = fs.readFileSync(entry.receipt);
  const q = JSON.parse(raw), record = JSON.parse(receipt);
  need(util.isDeepStrictEqual(q, record.request), 'request-receipt-mismatch');
  const envelope = record.response?.result, s = envelope?.structuredContent;
  need(envelope?.isError !== true && s?.ok === true, 'receipt-not-success');
  if (entry.kind === 'inspect') {
    need(Array.isArray(q.requests) && q.requests.every(x =>
      ['outline','match'].includes(x.operation) && typeof x.file === 'string'), 'unsupported-inspect-shape');
    const files = new Set(q.requests.map(x => x.file)).size;
    need(q.expect?.requests === q.requests.length && q.expect?.files === files, 'request-cardinality-mismatch');
    need(s.read_complete === true && s.request_count === q.requests.length && s.file_count === files, 'receipt-cardinality-mismatch');
  } else if (entry.kind === 'fanout') {
    need(Array.isArray(q.edits) && q.edits.every(x => typeof x.file === 'string' &&
      Number.isInteger(x.matches) && x.matches > 0), 'unsupported-fanout-shape');
    need(s.committed === true && s.files === new Set(q.edits.map(x => x.file)).size &&
      s.edits === q.edits.reduce((n,x) => n+x.matches,0), 'receipt-cardinality-mismatch');
  } else if (entry.kind === 'alias') {
    need(q.op === 'alias_migration' && Number.isInteger(q.expect?.files), 'unsupported-alias-shape');
    need(s.committed === true && s.files === q.expect.files, 'receipt-cardinality-mismatch');
  } else throw Error('unsupported-kind');
  return '## '+entry.kind+'\n\n```json\n'+raw.toString()+'```\n\n'+
    'Archived successful execution; request SHA256 '+hash(raw)+'; receipt SHA256 '+hash(receipt)+'.\n'+
    'This validates the retained specimen, not current source freshness, replay safety, or task semantics.\n';
}
try {
  const entries = JSON.parse(fs.readFileSync(process.argv[2]));
  // Validate all specimens before emitting any partial document.
  process.stdout.write(entries.map(specimen).join('\n'));
} catch (e) { process.stderr.write('REFUSED '+e.message+'\n'); process.exitCode=2; }

File: docs/observations/2026-09-06-astra-executable-specimen/specimen.clj
;; Same bounded prototype over Clojure data; no eval or request execution.
(require '[cheshire.core :as json])
(import '[java.security MessageDigest])

(defn sha [s]
  (format "%064x" (BigInteger. 1 (.digest (MessageDigest/getInstance "SHA-256")
                                          (.getBytes s "UTF-8")))))
(defn need [ok reason]
  (when-not ok (throw (ex-info reason {}))))
(defn specimen [{:keys [kind request receipt]}]
  (let [raw (slurp request) proof (slurp receipt)
        q (json/parse-string raw true) record (json/parse-string proof true)
        envelope (get-in record [:response :result]) s (:structuredContent envelope)]
    (need (= q (:request record)) "request-receipt-mismatch")
    (need (and (not (true? (:isError envelope))) (true? (:ok s))) "receipt-not-success")
    (case kind
      "inspect"
      (let [requests (:requests q) files (count (set (map :file requests)))]
        (need (and (vector? requests)
                   (every? #(and (#{"outline" "match"} (:operation %))
                                  (string? (:file %))) requests)) "unsupported-inspect-shape")
        (need (and (= (get-in q [:expect :requests]) (count requests))
                   (= (get-in q [:expect :files]) files)) "request-cardinality-mismatch")
        (need (and (true? (:read_complete s)) (= (:request_count s) (count requests))
                   (= (:file_count s) files)) "receipt-cardinality-mismatch"))
      "fanout"
      (let [edits (:edits q)]
        (need (and (vector? edits) (every? #(and (string? (:file %))
                   (integer? (:matches %)) (pos? (:matches %))) edits)) "unsupported-fanout-shape")
        (need (and (true? (:committed s)) (= (:files s) (count (set (map :file edits))))
                   (= (:edits s) (reduce + (map :matches edits)))) "receipt-cardinality-mismatch"))
      "alias"
      (do (need (and (= "alias_migration" (:op q)) (integer? (get-in q [:expect :files])))
                "unsupported-alias-shape")
          (need (and (true? (:committed s)) (= (:files s) (get-in q [:expect :files])))
                "receipt-cardinality-mismatch"))
      (throw (ex-info "unsupported-kind" {})))
    (str "## " kind "\n\n```json\n" raw "```\n\n"
         "Archived successful execution; request SHA256 " (sha raw)
         "; receipt SHA256 " (sha proof) ".\n"
         "This validates the retained specimen, not current source freshness, replay safety, or task semantics.\n")))
(try
  (let [entries (json/parse-string (slurp (first *command-line-args*)) true)
        ;; Realize before printing: a refused batch emits no partial document.
        documents (mapv specimen entries)]
    (print (clojure.string/join "\n" documents)))
  (catch Exception e (binding [*out* *err*] (println "REFUSED" (.getMessage e)))
         (System/exit 2)))
